p2p/raintree/peers_manager.go

```diff
@@ -0,0 +1,182 @@
+)
+
+// peersManager is in charge of handling operations on peers (like adding/removing them) within an AddrBook
+type peersManager struct {
+	selfAddr cryptoPocket.Address
+	eventCh  chan addressBookEvent
+
+	m  sync.RWMutex
+	wg sync.WaitGroup
+
+	addrBook typesP2P.AddrBook
+
+	addrBookMap  typesP2P.AddrBookMap
+	addrList     typesP2P.AddrList
+	maxNumLevels uint32
+}
+
+func newPeersManager(selfAddr cryptoPocket.Address, addrBook typesP2P.AddrBook) (*peersManager, error) {
+	pm := &peersManager{
+		selfAddr:     selfAddr,
+		addrBook:     addrBook,
+		eventCh:      make(chan addressBookEvent, 1),
+		addrBookMap:  make(typesP2P.AddrBookMap),
+		addrList:     make([]string, 0),
+		maxNumLevels: 0,
+	}
+
+	// inizializing map and list
+	pm.addrBookMap = make(map[string]*typesP2P.NetworkPeer, len(addrBook))
+	pm.addrList = make([]string, len(addrBook))
+	for i, peer := range addrBook {
+		addr := peer.Address.String()
+		pm.addrList[i] = addr
+		pm.addrBookMap[addr] = peer
+	}
+
+	sort.Strings(pm.addrList)
+
+	i := sort.SearchStrings(pm.addrList, pm.selfAddr.String())
+	if i == len(pm.addrList) {
+		return nil, fmt.Errorf("self address not found for %s in addrBook so this client can send messages but does not propagate them", pm.selfAddr)
+	}
+	// The list is sorted lexicographically above, but is reformatted below so this addr of this node
+	// is always the first in the list. This makes RainTree propagation easier to compute and interpret.
+	pm.addrList = append(pm.addrList[i:len(pm.addrList)], pm.addrList[0:i]...)
+
+	fmt.Printf("[DEBUG] pm.addrList post sort: %v\n", pm.addrList)
+
+	// sorting pm.addrBook as well, leveraging the sort order we just achieved
+	for i := 0; i < len(pm.addrList); i++ {
+		pm.addrBook[i] = pm.addrBookMap[pm.addrList[i]]
+	}
+
+	pm.maxNumLevels = pm.getMaxAddrBookLevels()
+
+	// listening and reacting to events
+	go func() {
+		for evt := range pm.eventCh {
+			pm.m.Lock()
+
+			peerAddress := evt.peer.Address.String()
+
+			switch evt.eventType {
+			case addToAddressBook:
+
+				pm.addrBookMap[peerAddress] = evt.peer
+				// insert into sorted addrList and addrBook
+				// searching from index 1 because index 0 is self by convention and the rest of the slice is sorted
+				i := sort.SearchStrings(pm.addrList[1:], peerAddress)
+				pm.addrList = insertElementAtIndex(pm.addrList, peerAddress, i)
+				pm.addrBook = insertElementAtIndex(pm.addrBook, evt.peer, i)
+
+				updateMaxNumLevels(pm)
+
+				pm.wg.Done()
+			case removeFromAddressBook:
+				delete(pm.addrBookMap, peerAddress)
+
+				// remove from sorted addrList and addrBook
+				// searching from index 1 because index 0 is self by convention and the rest of the slice is sorted
+				i := sort.SearchStrings(pm.addrList[1:], peerAddress)
+				pm.addrList = removeElementAtIndex(pm.addrList, i)
+				pm.addrBook = removeElementAtIndex(pm.addrBook, i)
+
+				updateMaxNumLevels(pm)
+
+				pm.wg.Done()
+			}
+
+			pm.m.Unlock()
```

Comment:
I will build out a small program to "prove my point to myself", but I find it weird that we have a mutex, a waitground AND a channel of size 1 even though golang channels of buffer size 1 can be used to block: https://stackoverflow.com/a/11943866/768439.

Will follow up on this tomorrow.

Reply:
Nvm. I understand.

We make sure that the event loop is both isolated and atomic, and that addition/removal is a blocking operation.

💯